src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java

```diff
@@ -0,0 +1,145 @@
+import software.amazon.awssdk.services.s3.model.ListObjectsRequest;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+import java.time.Clock;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * <p>
+ * Service that periodically calls s3 for stored request updates.
+ * If refreshRate is negative, then the data will never be refreshed.
+ * <p>
+ * Fetches all files from the specified folders/prefixes in s3 and downloads all files.
+ */
+public class S3PeriodicRefreshService implements Initializable {
+
+    private static final String JSON_SUFFIX = ".json";
+
+    private static final Logger logger = LoggerFactory.getLogger(S3PeriodicRefreshService.class);
+
+    private final S3AsyncClient asyncClient;
+    private final String bucket;
+    private final String storedRequestsDirectory;
+    private final String storedImpressionsDirectory;
+    private final long refreshPeriod;
+    private final CacheNotificationListener cacheNotificationListener;
+    private final MetricName cacheType;
+    private final Clock clock;
+    private final Metrics metrics;
+    private final Vertx vertx;
+
+    public S3PeriodicRefreshService(S3AsyncClient asyncClient,
+                                    String bucket,
+                                    String storedRequestsDirectory,
+                                    String storedImpressionsDirectory,
+                                    long refreshPeriod,
+                                    CacheNotificationListener cacheNotificationListener,
+                                    MetricName cacheType,
+                                    Clock clock,
+                                    Metrics metrics,
+                                    Vertx vertx) {
+
+        this.asyncClient = Objects.requireNonNull(asyncClient);
+        this.bucket = Objects.requireNonNull(bucket);
+        this.storedRequestsDirectory = Objects.requireNonNull(storedRequestsDirectory);
+        this.storedImpressionsDirectory = Objects.requireNonNull(storedImpressionsDirectory);
+        this.refreshPeriod = refreshPeriod;
+        this.cacheNotificationListener = Objects.requireNonNull(cacheNotificationListener);
+        this.cacheType = Objects.requireNonNull(cacheType);
+        this.clock = Objects.requireNonNull(clock);
+        this.metrics = Objects.requireNonNull(metrics);
+        this.vertx = Objects.requireNonNull(vertx);
+    }
+
+    @Override
+    public void initialize(Promise<Void> initializePromise) {
+        fetchStoredDataResult(clock.millis(), MetricName.initialize);
+        if (refreshPeriod > 0) {
+            logger.info("Starting s3 periodic refresh for " + cacheType + " every " + refreshPeriod + " s");
+            vertx.setPeriodic(refreshPeriod, ignored -> fetchStoredDataResult(clock.millis(), MetricName.update));
+        }
+
+        initializePromise.tryComplete();
```

Comment:
I'm still favouring the solution where the promise is completed with the first fetch result.

Failing to fetch stored impressions/requests is a critical sign that the new instance shouldn't startup. 

And it should also open a port until the initial data is loaded.